File: cathode/src/main/java/net/simonvt/cathode/ui/lists/ListAdapter.java
/*
 * Copyright (C) 2016 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.ui.lists;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import net.simonvt.cathode.R;
import net.simonvt.cathode.common.ui.adapter.BaseAdapter;
import net.simonvt.cathode.common.widget.OverflowView;
import net.simonvt.cathode.common.widget.RemoteImageView;
import net.simonvt.cathode.entity.ListEpisode;
import net.simonvt.cathode.entity.ListItem;
import net.simonvt.cathode.entity.ListMovie;
import net.simonvt.cathode.entity.ListPerson;
import net.simonvt.cathode.entity.ListSeason;
import net.simonvt.cathode.entity.ListShow;
import net.simonvt.cathode.images.ImageType;
import net.simonvt.cathode.images.ImageUri;
import net.simonvt.cathode.provider.util.DataHelper;

public class ListAdapter extends BaseAdapter<ListItem, ListAdapter.ListViewHolder> {

  interface ListListener {

    void onShowClick(long showId, String title, String overview);

    void onSeasonClick(long showId, long seasonId, String showTitle, int seasonNumber);

    void onEpisodeClick(long id);

    void onMovieClicked(long movieId, String title, String overview);

    void onPersonClick(long personId);

    void onRemoveItem(int position, ListItem listItem);
  }

  private static final int TYPE_SHOW = 1;
  private static final int TYPE_SEASON = 2;
  private static final int TYPE_EPISODE = 3;
  private static final int TYPE_MOVIE = 4;
  private static final int TYPE_PERSON = 5;

  ListListener listener;

  public ListAdapter(Context context, ListListener listener) {
    super(context);
    this.listener = listener;
  }

  @Override
  protected boolean areItemsTheSame(@NonNull ListItem oldItem, @NonNull ListItem newItem) {
    return oldItem.getListItemId() == newItem.getListItemId();
  }

  @Override public int getItemViewType(int position) {
    ListItem item = getList().get(position);
    switch (item.getType()) {
      case SHOW:
        return TYPE_SHOW;
      case SEASON:
        return TYPE_SEASON;
      case EPISODE:
        return TYPE_EPISODE;
      case MOVIE:
        return TYPE_MOVIE;
      case PERSON:
        return TYPE_PERSON;
      default:
        throw new IllegalStateException("Unsupported item type " + item.getType().toString());
    }
  }

  @Override public ListViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    ListViewHolder holder;

    if (viewType == TYPE_SHOW) {
      View v = LayoutInflater.from(getContext()).inflate(R.layout.row_list_show, parent, false);
      final ShowViewHolder showHolder = new ShowViewHolder(v);
      holder = showHolder;

      v.setOnClickListener(new View.OnClickListener() {
        @Override public void onClick(View v) {
          final int position = showHolder.getAdapterPosition();
          if (position != RecyclerView.NO_POSITION) {
            ListItem item = getList().get(position);
            ListShow show = item.getShow();
            listener.onShowClick(show.getId(), show.getTitle(), show.getOverview());
          }
        }
      });
    } else if (viewType == TYPE_SEASON) {
      View v = LayoutInflater.from(getContext()).inflate(R.layout.row_list_season, parent, false);
      final SeasonViewHolder seasonHolder = new SeasonViewHolder(v);
      holder = seasonHolder;

      v.setOnClickListener(new View.OnClickListener() {
        @Override public void onClick(View v) {
          final int position = seasonHolder.getAdapterPosition();
          if (position != RecyclerView.NO_POSITION) {
            ListItem item = getList().get(position);
            ListSeason season = item.getSeason();
            listener.onSeasonClick(season.getShowId(), season.getId(), season.getShowTitle(),
                season.getSeason());
          }
        }
      });
    } else if (viewType == TYPE_EPISODE) {
      View v = LayoutInflater.from(getContext()).inflate(R.layout.row_list_episode, parent, false);

      final EpisodeViewHolder episodeHolder = new EpisodeViewHolder(v);
      holder = episodeHolder;

      v.setOnClickListener(new View.OnClickListener() {
        @Override public void onClick(View v) {
          final int position = episodeHolder.getAdapterPosition();
          if (position != RecyclerView.NO_POSITION) {
            ListItem item = getList().get(position);
            listener.onEpisodeClick(item.getEpisode().getId());
          }
        }
      });
    } else if (viewType == TYPE_MOVIE) {
      View v = LayoutInflater.from(getContext()).inflate(R.layout.row_list_movie, parent, false);
      final MovieViewHolder movieHolder = new MovieViewHolder(v);
      holder = movieHolder;

      v.setOnClickListener(new View.OnClickListener() {
        @Override public void onClick(View v) {
          final int position = movieHolder.getAdapterPosition();
          if (position != RecyclerView.NO_POSITION) {
            ListItem item = getList().get(position);
            ListMovie movie = item.getMovie();
            listener.onMovieClicked(movie.getId(), movie.getTitle(), movie.getOverview());
          }
        }
      });
    } else {
      View v = LayoutInflater.from(getContext()).inflate(R.layout.row_list_person, parent, false);
      final PersonViewHolder personHolder = new PersonViewHolder(v);
      holder = personHolder;

      v.setOnClickListener(new View.OnClickListener() {
        @Override public void onClick(View v) {
          final int position = personHolder.getAdapterPosition();
          if (position != RecyclerView.NO_POSITION) {
            ListItem item = getList().get(position);
            listener.onPersonClick(item.getPerson().getId());
          }
        }
      });
    }

    final ListViewHolder finalHolder = holder;
    holder.overflow.addItem(R.id.action_list_remove, R.string.action_list_remove);
    holder.overflow.setListener(new OverflowView.OverflowActionListener() {
      @Override public void onPopupShown() {
      }

      @Override public void onPopupDismissed() {
      }

      @Override public void onActionSelected(int action) {
        final int position = finalHolder.getAdapterPosition();
        if (position != RecyclerView.NO_POSITION) {
          ListItem item = getList().get(position);
          switch (action) {
            case R.id.action_list_remove:
              listener.onRemoveItem(position, item);
              break;
          }
        }
      }
    });

    return holder;
  }

  @Override public void onViewRecycled(ListViewHolder holder) {
    holder.overflow.dismiss();
  }

  @Override public void onBindViewHolder(@NonNull ListViewHolder holder, int position) {
    ListItem item = getList().get(position);
    if (holder.getItemViewType() == TYPE_SHOW) {
      ShowViewHolder showHolder = (ShowViewHolder) holder;
      ListShow show = item.getShow();

      String poster = ImageUri.create(ImageUri.ITEM_SHOW, ImageType.POSTER, show.getId());
      showHolder.poster.setImage(poster);
      showHolder.title.setText(show.getTitle());
      showHolder.overview.setText(show.getOverview());
    } else if (holder.getItemViewType() == TYPE_SEASON) {
      SeasonViewHolder seasonHolder = (SeasonViewHolder) holder;
      ListSeason season = item.getSeason();

      String showPoster = ImageUri.create(ImageUri.ITEM_SHOW, ImageType.POSTER, season.getShowId());
      seasonHolder.poster.setImage(showPoster);
      seasonHolder.season.setText(
          getContext().getResources().getString(R.string.season_x, season.getSeason()));
      seasonHolder.show.setText(season.getShowTitle());
    } else if (holder.getItemViewType() == TYPE_EPISODE) {
      EpisodeViewHolder episodeHolder = (EpisodeViewHolder) holder;
      ListEpisode episode = item.getEpisode();

      String title =
          DataHelper.getEpisodeTitle(getContext(), episode.getTitle(), episode.getSeason(),
              episode.getEpisode(), episode.getWatched());
      String screenshotUri =
          ImageUri.create(ImageUri.ITEM_EPISODE, ImageType.STILL, episode.getId());
      episodeHolder.screen.setImage(screenshotUri);
      episodeHolder.title.setText(title);
      episodeHolder.showTitle.setText(episode.getShowTitle());
    } else if (holder.getItemViewType() == TYPE_MOVIE) {
      MovieViewHolder movieHolder = (MovieViewHolder) holder;
      ListMovie movie = item.getMovie();

      String poster = ImageUri.create(ImageUri.ITEM_MOVIE, ImageType.POSTER, movie.getId());
      movieHolder.poster.setImage(poster);
      movieHolder.title.setText(movie.getTitle());
      movieHolder.overview.setText(movie.getOverview());
    } else {
      PersonViewHolder personHolder = (PersonViewHolder) holder;
      ListPerson person = item.getPerson();

      String headshot = ImageUri.create(ImageUri.ITEM_PERSON, ImageType.PROFILE, person.getId());
      personHolder.headshot.setImage(headshot);
      personHolder.name.setText(person.getName());
    }
  }

  public static class ListViewHolder extends RecyclerView.ViewHolder {

    OverflowView overflow;

    public ListViewHolder(View v) {
      super(v);
      overflow = v.findViewById(R.id.overflow);
    }
  }

  public static class ShowViewHolder extends ListViewHolder {

    RemoteImageView poster;
    TextView title;
    TextView overview;

    public ShowViewHolder(View v) {
      super(v);
      poster = v.findViewById(R.id.poster);
      title = v.findViewById(R.id.title);
      overview = v.findViewById(R.id.overview);
    }
  }

  public static class SeasonViewHolder extends ListViewHolder {

    RemoteImageView poster;
    TextView season;
    TextView show;

    public SeasonViewHolder(View v) {
      super(v);
      poster = v.findViewById(R.id.poster);
      season = v.findViewById(R.id.season);
      show = v.findViewById(R.id.show);
    }
  }

  public static class EpisodeViewHolder extends ListViewHolder {

    RemoteImageView screen;
    TextView title;
    TextView showTitle;

    EpisodeViewHolder(View v) {
      super(v);
      screen = v.findViewById(R.id.screen);
      title = v.findViewById(R.id.title);
      showTitle = v.findViewById(R.id.showTitle);
    }
  }

  public static class MovieViewHolder extends ListViewHolder {

    public RemoteImageView poster;
    public TextView title;
    public TextView overview;

    public MovieViewHolder(View v) {
      super(v);
      poster = v.findViewById(R.id.poster);
      title = v.findViewById(R.id.title);
      overview = v.findViewById(R.id.overview);
    }
  }

  public static class PersonViewHolder extends ListViewHolder {

    RemoteImageView headshot;
    TextView name;

    public PersonViewHolder(View v) {
      super(v);
      headshot = v.findViewById(R.id.headshot);
      name = v.findViewById(R.id.person_name);
    }
  }
}

File: cathode/src/main/java/net/simonvt/cathode/ui/lists/ListAdapter.kt
/*
 * Copyright (C) 2016 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.ui.lists

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import net.simonvt.cathode.R
import net.simonvt.cathode.api.enumeration.ItemType.EPISODE
import net.simonvt.cathode.api.enumeration.ItemType.MOVIE
import net.simonvt.cathode.api.enumeration.ItemType.PERSON
import net.simonvt.cathode.api.enumeration.ItemType.SEASON
import net.simonvt.cathode.api.enumeration.ItemType.SHOW
import net.simonvt.cathode.common.ui.adapter.BaseAdapter
import net.simonvt.cathode.common.widget.OverflowView
import net.simonvt.cathode.common.widget.OverflowView.OverflowActionListener
import net.simonvt.cathode.databinding.RowListEpisodeBinding
import net.simonvt.cathode.databinding.RowListMovieBinding
import net.simonvt.cathode.databinding.RowListPersonBinding
import net.simonvt.cathode.databinding.RowListSeasonBinding
import net.simonvt.cathode.databinding.RowListShowBinding
import net.simonvt.cathode.entity.ListItem
import net.simonvt.cathode.images.ImageType.POSTER
import net.simonvt.cathode.images.ImageType.PROFILE
import net.simonvt.cathode.images.ImageType.STILL
import net.simonvt.cathode.images.ImageUri
import net.simonvt.cathode.provider.util.DataHelper
import net.simonvt.cathode.ui.lists.ListAdapter.ListViewHolder

class ListAdapter(context: Context?, var listener: ListListener) :
  BaseAdapter<ListItem?, ListViewHolder?>(context) {

  interface ListListener {
    fun onShowClick(showId: Long, title: String?, overview: String?)
    fun onSeasonClick(showId: Long, seasonId: Long, showTitle: String?, seasonNumber: Int)
    fun onEpisodeClick(id: Long)
    fun onMovieClicked(movieId: Long, title: String?, overview: String?)
    fun onPersonClick(personId: Long)
    fun onRemoveItem(position: Int, listItem: ListItem)
  }

  override fun areItemsTheSame(oldItem: ListItem, newItem: ListItem): Boolean {
    return oldItem.listItemId == newItem.listItemId
  }

  override fun getItemViewType(position: Int): Int {
    val item = list[position]!!
    return when (item.type) {
      SHOW -> TYPE_SHOW
      SEASON -> TYPE_SEASON
      EPISODE -> TYPE_EPISODE
      MOVIE -> TYPE_MOVIE
      PERSON -> TYPE_PERSON
      else -> throw IllegalStateException("Unsupported item type ${item.type}")
    }
  }

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListViewHolder {
    val holder: ListViewHolder
    if (viewType == TYPE_SHOW) {
      val binding = RowListShowBinding.inflate(LayoutInflater.from(context), parent, false)
      val showHolder = ShowViewHolder(binding)
      holder = showHolder
      binding.root.setOnClickListener {
        val position = showHolder.adapterPosition
        if (position != RecyclerView.NO_POSITION) {
          val show = list[position]!!.show!!
          listener.onShowClick(show.id, show.title, show.overview)
        }
      }
    } else if (viewType == TYPE_SEASON) {
      val binding = RowListSeasonBinding.inflate(LayoutInflater.from(context), parent, false)
      val seasonHolder = SeasonViewHolder(binding)
      holder = seasonHolder
      binding.root.setOnClickListener {
        val position = seasonHolder.adapterPosition
        if (position != RecyclerView.NO_POSITION) {
          val season = list[position]!!.season!!
          listener.onSeasonClick(season.showId, season.id, season.showTitle, season.season)
        }
      }
    } else if (viewType == TYPE_EPISODE) {
      val binding = RowListEpisodeBinding.inflate(LayoutInflater.from(context), parent, false)
      val episodeHolder = EpisodeViewHolder(binding)
      holder = episodeHolder
      binding.root.setOnClickListener {
        val position = episodeHolder.adapterPosition
        if (position != RecyclerView.NO_POSITION) {
          val episode = list[position]!!.episode!!
          listener.onEpisodeClick(episode.id)
        }
      }
    } else if (viewType == TYPE_MOVIE) {
      val binding = RowListMovieBinding.inflate(LayoutInflater.from(context), parent, false)
      val movieHolder = MovieViewHolder(binding)
      holder = movieHolder
      binding.root.setOnClickListener {
        val position = movieHolder.adapterPosition
        if (position != RecyclerView.NO_POSITION) {
          val movie = list[position]!!.movie!!
          listener.onMovieClicked(movie.id, movie.title, movie.overview)
        }
      }
    } else {
      val binding = RowListPersonBinding.inflate(LayoutInflater.from(context), parent, false)
      val personHolder = PersonViewHolder(binding)
      holder = personHolder
      binding.root.setOnClickListener {
        val position = personHolder.adapterPosition
        if (position != RecyclerView.NO_POSITION) {
          val item = list[position]!!
          listener.onPersonClick(item.person!!.id)
        }
      }
    }
    holder.overflow.addItem(R.id.action_list_remove, R.string.action_list_remove)
    holder.overflow.setListener(object : OverflowActionListener {
      override fun onPopupShown() {}
      override fun onPopupDismissed() {}
      override fun onActionSelected(action: Int) {
        val position = holder.adapterPosition
        if (position != RecyclerView.NO_POSITION) {
          val item = list[position]!!
          when (action) {
            R.id.action_list_remove -> listener.onRemoveItem(position, item)
          }
        }
      }
    })
    return holder
  }

  override fun onViewRecycled(holder: ListViewHolder) {
    holder.overflow.dismiss()
  }

  override fun onBindViewHolder(holder: ListViewHolder, position: Int) {
    val (_, _, _, _, _, show, season, episode, movie, person) = list[position]!!
    if (holder.itemViewType == TYPE_SHOW) {
      val showHolder = holder as ShowViewHolder
      val poster = ImageUri.create(ImageUri.ITEM_SHOW, POSTER, show!!.id)
      showHolder.poster.setImage(poster)
      showHolder.title.text = show.title
      showHolder.overview.text = show.overview
    } else if (holder.itemViewType == TYPE_SEASON) {
      val seasonHolder = holder as SeasonViewHolder
      val showPoster = ImageUri.create(ImageUri.ITEM_SHOW, POSTER, season!!.showId)
      seasonHolder.poster.setImage(showPoster)
      seasonHolder.season.text = context.resources.getString(R.string.season_x, season.season)
      seasonHolder.show.text = season.showTitle
    } else if (holder.itemViewType == TYPE_EPISODE) {
      val episodeHolder = holder as EpisodeViewHolder
      val title = DataHelper.getEpisodeTitle(
        context,
        episode!!.title,
        episode.season,
        episode.episode,
        episode.watched
      )
      val screenshotUri = ImageUri.create(ImageUri.ITEM_EPISODE, STILL, episode.id)
      episodeHolder.screen.setImage(screenshotUri)
      episodeHolder.title.text = title
      episodeHolder.showTitle.text = episode.showTitle
    } else if (holder.itemViewType == TYPE_MOVIE) {
      val movieHolder = holder as MovieViewHolder
      val poster = ImageUri.create(ImageUri.ITEM_MOVIE, POSTER, movie!!.id)
      movieHolder.poster.setImage(poster)
      movieHolder.title.text = movie.title
      movieHolder.overview.text = movie.overview
    } else {
      val personHolder = holder as PersonViewHolder
      val headshot = ImageUri.create(ImageUri.ITEM_PERSON, PROFILE, person!!.id)
      personHolder.headshot.setImage(headshot)
      personHolder.name.text = person.name
    }
  }

  open class ListViewHolder(v: View, val overflow: OverflowView) : ViewHolder(v)

  class ShowViewHolder(binding: RowListShowBinding) :
    ListViewHolder(binding.root, binding.overflow) {
    val poster = binding.poster
    val title = binding.title
    val overview = binding.overview
  }

  class SeasonViewHolder(binding: RowListSeasonBinding) :
    ListViewHolder(binding.root, binding.overflow) {
    val poster = binding.poster
    val season = binding.season
    val show = binding.show
  }

  class EpisodeViewHolder(binding: RowListEpisodeBinding) :
    ListViewHolder(binding.root, binding.overflow) {
    val screen = binding.screen
    val title = binding.title
    val showTitle = binding.showTitle
  }

  class MovieViewHolder(binding: RowListMovieBinding) :
    ListViewHolder(binding.root, binding.overflow) {
    val poster = binding.poster
    val title = binding.title
    val overview = binding.overview
  }

  class PersonViewHolder(binding: RowListPersonBinding) :
    ListViewHolder(binding.root, binding.overflow) {
    val headshot = binding.headshot
    val name = binding.personName
  }

  companion object {
    private const val TYPE_SHOW = 1
    private const val TYPE_SEASON = 2
    private const val TYPE_EPISODE = 3
    private const val TYPE_MOVIE = 4
    private const val TYPE_PERSON = 5
  }
}
